### PR TITLE
fix: Remove non-authorized cat options from query [DHIS2-7712]

### PR DESCRIPTION
**Pull request by maikelarabori**
Inverting the logic so left not authorized category options are filtered out during the SQL query.

### CODE REVIEW

**Comment by maikelarabori** [comment]
Reopening due to this error: security/snyk - dhis-2/dhis-services/dhis-service-program-rule/pom.xml

Trying again.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java [hunk 1]
```diff
@@ -520,18 +520,17 @@ else if ( params.hasProgram() && params.getProgram().hasCategoryCombo() )
     }
 
     /**
-     * This method will generate a sql sentence responsible for filtering only the
-     * allowed category options for this program, and expects to receive only the
-     * authorized category options.
+     * This method will generate a sql sentence responsible for filtering out all
+     * the category options (of this program) not authorized.
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java]: Can you elaborate on the javadoc?  not authorized in which sense?

**Comment by maikelarabori** [comment]
Reopening due to Snyk issues.

**Comment by stian-sandvold** [comment]
Code looks good. Just a thought: Would it be safer to do the data read check as part of the sql, instead of injecting (potentially a very large number) uids into the "not in" sql?

**Comment by maikelarabori** [comment]
> Code looks good. Just a thought: Would it be safer to do the data read check as part of the sql, instead of injecting (potentially a very large number) uids into the "not in" sql?

Hi @stian-sandvold,
that could be done as well, but then we would end up with these situations:

1) we would need to bring the authorization logic into the JDBC query level;
2) we would apply a different approach instead of the current one, where all authorization checks in analytics are being performed before the query.

I opted to concentrate the checkings in one place (as we have now) until we have a more generic component that we can use for authorization checkings.

Maybe I didn't get your full idea, so please let me know if you want to have a quick chat around it. :+1:

**Pull request merged**